server/src/main/java/org/candlepin/model/AbstractHibernateCurator.java

```diff
@@ -767,47 +770,205 @@ protected int safeSQLUpdateWithCollection(String sql, Collection<?> collection,
+            @Override
+            public Iterator<String> iterator() {
+                return new Iterator<String>() {
+                    private Iterator<E> entityIterator;
+
+                    /* initializer */ {
+                        this.entityIterator = entityCollection.iterator();
+                    }
+
+                    @Override
+                    public boolean hasNext() {
+                        return this.entityIterator.hasNext();
+                    }
+
+                    @Override
+                    public String next() {
+                        E next = this.entityIterator.next();
+                        return next != null ? (String) next.getId() : null;
+                    }
+
+                    @Override
+                    public void remove() {
+                        this.entityIterator.remove();
+                    }
+                };
+            }
+        };
+
+        return this.lockAndLoadByColumn(this.entityType.getSimpleName(), "id", iterable);
     }
 
-    public List<E> lockAndLoadBatchById(Iterable<String> ids) {
-        List<E> result = new LinkedList<E>();
-        for (String id : ids) {
-            // Load at the current lock level so an existing object in the session will be loaded
-            // if it already exists.
-            E entity = getEntityManager().find(this.entityType(), id);
-            // Use refresh in order to upgrade the lock level & return the new version of the entity
-            getEntityManager().refresh(entity, LockModeType.PESSIMISTIC_WRITE);
-            result.add(entity);
+    /**
+     * Loads the entities represented by the given IDs with a pessimistic lock. If no entities were
+     * found with the given IDs, this method returns an empty collection.
+     *
+     * @param ids
+     *  A collection of entity IDs to use to load and lock the represented entities
+     *
+     * @return
+     *  A collection of locked entities represented by the given IDs
+     */
+    public Collection<E> lockAndLoadByIds(Iterable<String> ids) {
+        return this.lockAndLoadByColumn(this.entityType.getSimpleName(), "id", ids);
+    }
+
+    /**
+     * Loads an entity with a pessimistic lock using the specified key name and value. If a
+     * matching entity could not be found, this method returns null.
+     * <p></p>
+     * Note that while the key name could refer to a non-index column, doing so may result in an
+     * exception if multiple entities are matched.
+     *
+     * @param entityName
+     *  The name of the entity to load (i.e. Pool or Product)
+     *
+     * @param keyName
+     *  The name of the indexed column/property to examine (i.e. id or key)
+     *
+     * @param value
+     *  The value the specified key must match
+     *
+     * @return
+     *  A locked entity instance, or null if a matching entity could not be found
+     */
+    protected E lockAndLoadByKey(String entityName, String keyName, Object value) {
```

Comment:
Unless I'm overlooking something, it doesn't look like this will be an option here afterall. Switching over to TypedQuery gives us no control over the level 2 cache mode, which puts us one 
step closer to needing a full refresh on everything.